src/object/cli_ec.c

```diff
@@ -1502,6 +1663,434 @@ obj_ec_recov_prep(struct obj_reasb_req *reasb_req, daos_obj_id_t oid,
+			rtask->ert_iod.iod_type = iod->iod_type;
+			rtask->ert_iod.iod_size = iod->iod_size;
+			rtask->ert_iod.iod_nr = 1;
+			rtask->ert_iod.iod_recxs = &recx_ep->re_recx;
+			rtask->ert_epoch = recx_ep->re_ep;
+			rc = daos_sgl_init(&rtask->ert_sgl, 1);
+			if (rc)
+				goto out;
+			buf_sz = stripe_nr * stripe_total_sz;
+			D_ASSERT(buf_off + buf_sz <=
+				 sgl->sg_iovs[0].iov_buf_len);
+			d_iov_set(&rtask->ert_sgl.sg_iovs[0],
+				  sgl->sg_iovs[0].iov_buf + buf_off, buf_sz);
+			buf_off += buf_sz;
+		}
+	}
+
+out:
+	return rc;
+}
+
+void
+obj_ec_fail_info_free(struct obj_reasb_req *reasb_req)
+{
+	struct obj_ec_fail_info *fail_info = reasb_req->orr_fail;
+
+	if (fail_info == NULL || reasb_req->orr_recov == 1)
+		return;
+
+	obj_ec_recov_task_fini(reasb_req);
+	obj_ec_recov_codec_free(reasb_req);
+	daos_recx_ep_list_free(fail_info->efi_recx_lists,
+			       fail_info->efi_nrecx_lists);
+	daos_recx_ep_list_free(fail_info->efi_stripe_lists,
+			       fail_info->efi_nrecx_lists);
+	D_FREE(fail_info->efi_tgt_list);
+	D_FREE(fail_info);
+	reasb_req->orr_fail = NULL;
+}
+
+int
+obj_ec_recov_prep(struct obj_reasb_req *reasb_req, daos_obj_id_t oid,
+		  daos_iod_t *iods, uint32_t iod_nr)
+{
+	struct obj_ec_fail_info	*fail_info = reasb_req->orr_fail;
+	int			 rc;
+
+	D_ASSERT(iod_nr == fail_info->efi_nrecx_lists);
+	rc = obj_ec_stripe_list_init(fail_info->efi_recx_lists,
+				    fail_info->efi_nrecx_lists,
+				    reasb_req->orr_oca,
+				   &fail_info->efi_stripe_lists);
+	if (rc)
+		goto out;
+
+	rc = obj_ec_recov_task_init(reasb_req, oid, iods, iod_nr);
+	if (rc)
+		goto out;
+
+	rc = obj_ec_recov_codec_init(reasb_req, oid, fail_info->efi_ntgts,
+				     fail_info->efi_tgt_list);
+	if (rc)
+		goto out;
+
+out:
+	if (rc)
+		D_ERROR(DF_OID" obj_ec_recov_prep failed, "DF_RC".\n",
+			DP_OID(oid), DP_RC(rc));
+	return rc;
+}
+
+static void
+obj_ec_recov_stripe(struct obj_ec_recov_codec *codec,
+		    struct daos_oclass_attr *oca, void *buf_stripe,
+		    uint64_t cell_sz)
+{
+	unsigned char	*buf_src[OBJ_EC_MAX_K];
+	unsigned char	*buf_err[OBJ_EC_MAX_P];
+	uint32_t	 i, k;
+
+	k = obj_ec_data_tgt_nr(oca);
+	for (i = 0; i < k; i++)
+		buf_src[i] = buf_stripe + codec->er_dec_idx[i] * cell_sz;
+	for (i = 0; i < codec->er_nerrs; i++)
+		buf_err[i] = buf_stripe + codec->er_err_list[i] * cell_sz;
+
+	ec_encode_data(cell_sz, k, codec->er_nerrs, codec->er_gftbls,
+		       buf_src, buf_err);
+}
+
+struct oes_copy_arg {
+	void		*buf;
+	uint64_t	 size;
+	uint64_t	 copied;
+};
+
+static int
+oes_copy(uint8_t *buf, size_t len, void *data)
+{
+	struct oes_copy_arg	*arg = data;
+
+	D_ASSERT(arg->copied + len <= arg->size);
+	memcpy(buf, arg->buf + arg->copied, len);
+	arg->copied += len;
+
+	return 0;
+}
+
+static void
+obj_ec_sgl_copy(d_sg_list_t *sgl, uint64_t off, void *buf, uint64_t size)
+{
+	struct daos_sgl_idx	sgl_idx = {0};
+	struct oes_copy_arg	arg;
+	int			rc;
+
+	rc = daos_sgl_processor(sgl, &sgl_idx, off, NULL, NULL);
```

Comment:
is this noop? Could you please add some comment here?

Reply:
it is to skip sgl to off, i'll add some comments when refresh this patch, thx